internal/pkg/logging/task.go

```diff
@@ -117,6 +117,14 @@ func (t *TaskClient) allTasksStopped() (bool, error) {
 		}
 	}
 	t.tasks = runningTasks
+
+	if stopped {
```

Comment:
The `logging` package is used for writing stdout logs to an `io.Writer` from ECS tasks or an App Runner service.
I think this functionality should be moved to a different package to return the exit codes for the containers in a given task.

For example:
1. Creating the functionality in the `internal/pkg/task ` pkg:
```go
package task

// ContainerDescription represents a container part of an ECS task.
type ContainerDescription struct {
   Name string   // Friendly name for the container.
   TaskID string // The task ID the container belongs to. 
   
   ExitCode int64
   ExitReason string
}

// HumanString returns a friendly colorized summary of the container description.
func (cd *ContainerDescription) HumanString() string {
   buf := new(strings.Builder)
   code := fmt.Sprintf("%s", cd.ExitCode)
   if cd.ExitCode > 0 {
      code = color.Red.Sprint(cd.ExitCode)
   }
   buf.WriteString(fmt.Sprintf("Container %s in task %s exited with status code %s", cd.Name, cd.TaskID, code)
   if cd.ExitReason != "" {
      buf.WriteString(" wit reason: %s", cd.ExitReason)
   }
   return buf.String()
}

// DescribeContainers returns the status of the containers part of the given tasks.
func (r *EnvRunner) DescribeContainers(tasks []*Task) []Container { ... }

// DescribeContainers returns the status of the containers part of the given tasks.
func (r *ConfigRunner) DescribeContainers(tasks []*Task) []Container { ... }
```

2. Moving the logging functionality to the `cli` pkg:
   https://github.com/aws/copilot-cli/blob/81b3dbbe9dc8cc8c65ebf5b7bc01b2f098ea609f/internal/pkg/cli/task_run.go#L587-L592
```go
if o.follow {
   ... 
   containers, err := o.runner.DescribeContainers(tasks)
   if err != nil {
       return fmt.Errorf("describe containers in running tasks: %v", err)
   }
   for _, container := range containers {
       log.Println(container.HumanString())
   }
}
```

Reply:
Hi @efekarakus, 👋  sorry for the delay and thanks for your comments.

I understand that you require the new changes need to fit the existing architecture, sorry I didn't noticed.

I 've been juggling my local code based on your notes, I didn't actually run the modified code.. but, I couldn't get two things:

* First, how is `task_run.go` going to execute `containers, err := o.runner.DescribeContainers(tasks)`? Because `o.runner` is under the `Runner` interface which only exposes `RunTask` method, I know it is in fact rather an `EnvRunner` or a `ConfigRunner` which they will have the `DescribeContainers` defined, but how do you access it through the `Runner` interface? Do you downcast, need an extra finer grain interface? Maybe I 'm missing something related with go which I 'm not familiar with.

* Then on `task/task.go` I was trying to implement

```
func (r *EnvRunner) DescribeContainers(tasks []*Task) []Container { ... }
func (r *ConfigRunner) DescribeContainers(tasks []*Task) []Container { ... }
```
So I was thinking on calling `aws/ecs/ecs.go`'s `DescribeTasks` there, but I 'm not sure how to get a handle to an instantiated ecs client object from that context.

Reply:
Hi @nbarrera 👋 

Great questions, apologies for the delay 🙇  Please let me know if any of the clarifications below don't make sense:

> First, how is task_run.go going to execute containers, err := o.runner.DescribeContainers(tasks)? 

We can update the interface: https://github.com/aws/copilot-cli/blob/78d70410d345ac4c6d7b7fb808a235f526e3f3c9/internal/pkg/cli/interfaces.go#L389-L391 to something like:
```go
type taskDescribeRunner interface {
     Run() ([]*task.Task, error)
     DescribeContainers(tasks []*Task) []Container { ... }
}
```
Note that after making this change, you'll want to run `make gen-mocks` to regenerate the Go mock types.

> but I 'm not sure how to get a handle to an instantiated ecs client object from that context.

Both `EnvRunner` and `ConfigRunner` have a `Starter` field which is really an ECS service 😄: https://github.com/aws/copilot-cli/blob/78d70410d345ac4c6d7b7fb808a235f526e3f3c9/internal/pkg/task/env_runner.go#L45
My recommendations are the following:
1. Rename `Starter` to `ECS` since tasks are specific resources to ECS.
2. Rename the `Runner` interface in the `task` pkg to allow `DescribeTasks`:
```go
type ECSClient interface {
   RunTask(input ecs.RunTaskInput) ([]*ecs.Task, error)
   DescribeTasks(cluster string, taskARNs []string) ([]*ecs.Task, error) 
}
```
3. Run `make gen-mocks`

Hope this helps!